File: src/Services/Implementations/AuthService.cs
using System.Text.RegularExpressions;
using ECommerceApi.Authentication;
using ECommerceApi.DTOs.Auth;
using ECommerceApi.Exceptions;
using ECommerceApi.Models;
using ECommerceApi.Repositories.Interfaces;
using ECommerceApi.Services.Interfaces;

namespace ECommerceApi.Services.Implementations;

/*
  Business logic for registration/login. Deliberately has no dependency on HttpContext/ControllerBase and depends only on 
  IUserRepository (an interface, not a concrete EF Core-backed class) 
*/
public class AuthService : IAuthService
{
    private static readonly Regex EmailRegex = new(@"^[^@\s]+@[^@\s]+\.[^@\s]+$", RegexOptions.Compiled);

    private readonly IUserRepository _userRepository;
    private readonly IPasswordHasher _passwordHasher;
    private readonly ITokenService _tokenService;

    public AuthService(IUserRepository userRepository, IPasswordHasher passwordHasher, ITokenService tokenService)
    {
        _userRepository = userRepository;
        _passwordHasher = passwordHasher;
        _tokenService = tokenService;
    }

    public async Task<AuthResponse> RegisterAsync(RegisterRequest request)
    {
        ValidateRegistration(request);

        var existing = await _userRepository.FindByEmailAsync(request.Email);
        if (existing is not null)
        {
            throw new ConflictAppException("An account with this email already exists");
        }

        var passwordHash = _passwordHasher.Hash(request.Password);
        var user = await _userRepository.CreateAsync(request.Email, passwordHash, request.FullName);

        return BuildResult(user);
    }

    public async Task<AuthResponse> LoginAsync(LoginRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.Email) || string.IsNullOrWhiteSpace(request.Password))
        {
            throw new ValidationException("Email and password are required");
        }

        var user = await _userRepository.FindByEmailAsync(request.Email);

        // Deliberately identical error for "no such user" and "wrong
        // password". Doesn't let a client enumerate emails.
        if (user is null || !_passwordHasher.Verify(request.Password, user.PasswordHash))
        {
            throw new UnauthorizedAppException("Invalid email or password");
        }

        return BuildResult(user);
    }

    private static void ValidateRegistration(RegisterRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.Email) || !EmailRegex.IsMatch(request.Email))
        {
            throw new ValidationException("A valid email address is required");
        }

        if (string.IsNullOrWhiteSpace(request.FullName))
        {
            throw new ValidationException("Full name is required");
        }

        var password = request.Password ?? string.Empty;
        if (password.Length < 8)
        {
            throw new ValidationException("Password must be at least 8 characters");
        }
        if (!password.Any(char.IsLetter))
        {
            throw new ValidationException("Password must contain at least one letter");
        }
        if (!password.Any(char.IsDigit))
        {
            throw new ValidationException("Password must contain at least one number");
        }
    }

    private AuthResponse BuildResult(User user)
    {
        var token = _tokenService.GenerateToken(user.Id, user.Role);
        return new AuthResponse
        {
            Token = token,
            User = new UserResponse
            {
                Id = user.Id,
                Email = user.Email,
                FullName = user.FullName,
                Role = user.Role.ToString().ToLowerInvariant(),
                CreatedAt = user.CreatedAt
            }
        };
    }
}

File: src/services/auth.service.ts
import { IUserRepository } from '../repositories/interfaces';
import { ConflictError, UnauthorizedError, ValidationError } from '../errors/AppError';
import { hashPassword, comparePassword } from '../utils/password';
import { signToken } from '../utils/jwt';
import { registerSchema, loginSchema } from '../validators/auth.validators';
import { SafeUser } from '../types/auth';
import { UserRecord } from '../types/domain';

export interface RegisterDTO {
  email: string;
  password: string;
  fullName: string;
}

export interface LoginDTO {
  email: string;
  password: string;
}

export interface AuthResult {
  user: SafeUser;
  token: string;
}

/**
 * Business logic for registration/login. Deliberately has no dependency on
 * Express request/response objects (Milestone 2 §4) and depends only on
 * the IUserRepository interface, not a concrete Prisma-backed class — so
 * it can be fully unit-tested with a mocked repository (Milestone 2 §23).
 */
export class AuthService {
  constructor(private readonly userRepository: IUserRepository) {}

  async register(input: RegisterDTO): Promise<AuthResult> {
    const parsed = registerSchema.safeParse(input);
    if (!parsed.success) {
      throw new ValidationError('Invalid registration input', parsed.error.flatten());
    }

    const existing = await this.userRepository.findByEmail(parsed.data.email);
    if (existing) {
      throw new ConflictError('An account with this email already exists');
    }

    const passwordHash = await hashPassword(parsed.data.password);
    const user = await this.userRepository.create({
      email: parsed.data.email,
      passwordHash,
      fullName: parsed.data.fullName
    });

    return this.buildResult(user);
  }

  async login(input: LoginDTO): Promise<AuthResult> {
    const parsed = loginSchema.safeParse(input);
    if (!parsed.success) {
      throw new ValidationError('Invalid login input', parsed.error.flatten());
    }

    const user = await this.userRepository.findByEmail(parsed.data.email);
    // Deliberately identical error/message for "no such user" and "wrong
    // password" (Milestone 2 §13) — do not let a client enumerate emails.
    if (!user) {
      throw new UnauthorizedError('Invalid email or password');
    }

    const passwordMatches = await comparePassword(parsed.data.password, user.passwordHash);
    if (!passwordMatches) {
      throw new UnauthorizedError('Invalid email or password');
    }

    return this.buildResult(user);
  }

  private buildResult(user: UserRecord): AuthResult {
    const token = signToken({ sub: user.id, role: user.role });
    const safeUser: SafeUser = {
      id: user.id,
      email: user.email,
      fullName: user.fullName,
      role: user.role,
      createdAt: user.createdAt
    };
    return { user: safeUser, token };
  }
}
